lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -103,51 +158,219 @@ TxIn
-    utxoTableWalletSlot   W.SlotId    sql=slot_id
+    -- The wallet checkpoint (wallet_id, slot)
+    utxoTableWalletId        W.WalletId  sql=wallet_id
+    utxoTableCheckpointSlot  W.SlotId    sql=slot
 
     -- TxIn
-    utxoTableInputId      TxId        sql=input_tx_id
-    utxoTableInputIndex   Word32      sql=input_index
+    utxoTableInputId         TxId        sql=input_tx_id
+    utxoTableInputIndex      Word32      sql=input_index
 
     -- TxOut
-    utxoTableOutputId     TxId        sql=output_tx_id
-    utxoTableOutputIndex  Word32      sql=output_index
+    utxoTableOutputAddress   W.Address   sql=output_address
+    utxoTableOutputCoin      W.Coin      sql=output_coin
 
     Primary
         utxoTableWalletId
-        utxoTableWalletSlot
+        utxoTableCheckpointSlot
         utxoTableInputId
         utxoTableInputIndex
-        utxoTableOutputId
-        utxoTableOutputIndex
+        utxoTableOutputAddress
+        utxoTableOutputCoin
 
-    Foreign Checkpoint fk_checkpoint_utxo utxoTableWalletId utxoTableWalletSlot
+    Foreign Checkpoint fk_checkpoint_utxo utxoTableWalletId utxoTableCheckpointSlot
     deriving Show Generic
 |]
+
+
+----------------------------------------------------------------------------
+-- Sqlite connection set up
+
+enableForeignKeys :: Sqlite.Connection -> IO ()
+enableForeignKeys conn = stmt >>= void . Sqlite.step
+    where stmt = Sqlite.prepare conn "PRAGMA foreign_keys = ON;"
+
+createSqliteBackend :: Maybe FilePath -> LogFunc -> IO SqlBackend
+createSqliteBackend fp logFunc = do
+  conn <- Sqlite.open (sqliteConnStr fp)
+  enableForeignKeys conn
+  wrapConnection conn logFunc
+
+sqliteConnStr :: Maybe FilePath -> Text
+sqliteConnStr = maybe ":memory:" T.pack
+
+logStderr :: LogFunc
+logStderr _ _ _ str = B8.hPutStrLn stderr (fromLogStr str)
+
+-- | Run a query without error handling. There will be exceptions thrown if it
+-- fails.
+runQuery
+    :: SqlBackend
+    -> SqlPersistM a
+    -> IO a
+runQuery conn = runResourceT . runNoLoggingT . flip runSqlConn conn
+
+-- | Run an action, and convert any Sqlite constraints exception into the given
+-- error result. No other exceptions are handled.
+handleConstraint :: MonadCatch m => e -> m a -> m (Either e a)
+handleConstraint e = handleJust select handler . fmap Right
+  where
+    select (SqliteException ErrorConstraint _ _) = Just ()
+    select _ = Nothing
+    handler = const . pure  . Left $ e
+
+----------------------------------------------------------------------------
+-- Database layer methods
+
+-- | Sets up a connection to the SQLite database.
+--
+-- Database migrations are run to create tables if necessary.
+--
+-- If the given file path does not exist, it will be created by the sqlite
+-- library.
+newDBLayer
+    :: Maybe FilePath
+       -- ^ Database file location, or Nothing for in-memory database
+    -> IO (DBLayer IO s t)
+newDBLayer fp = do
+    conn <- createSqliteBackend fp logStderr
```

Comment:
wil this `logStderr` enable all logs (CREATE_TABLE, INSERT, ...) as a default or errors only? I would give my votes for the second option

Reply:
indeed, it introduces logs like : 
```Cardano.Wallet.DB.Sqlite
  Wallet table
Migrating: CREATE TABLE "wallet"("wallet_id" VARCHAR NOT NULL,"name" VARCHAR NOT NULL,"passphrase_last_updated_at" TIMESTAMP NULL,"status" INTEGER NOT NULL,"delegation" VARCHAR NULL, PRIMARY KEY ("wallet_id"))
CREATE TABLE "wallet"("wallet_id" VARCHAR NOT NULL,"name" VARCHAR NOT NULL,"passphrase_last_updated_at" TIMESTAMP NULL,"status" INTEGER NOT NULL,"delegation" VARCHAR NULL, PRIMARY KEY ("wallet_id")); []
```

Reply:
added more customizable logging function (one can choose LogLevel)